### PR TITLE
initial pass on azure blob storage, filesystem storage + endpoint

### PR DESCRIPTION
**Pull request by DnOberon**
- Rewrote the `FileT` storage Create method to accept only a single payload, not an array. We don't need the bulk create functionality here and it's only going to cause bloated code down the line
- Created Azure Blob Storage implementation of the `FileStorage` interface. Implemented all members
- Created Filesystem implementation of the `FileStorage` interface. Implemented all members
- Tests on above implementations
- Created an endpoint for uploading files through a data source. Endpoint can take any number of files and additional form values. Additional form values are saved as an import with the file names injected as other properties.
- Modified the Mock implementation of `FileStorage` in order to implement the interface correctly
- Updated the Swagger and Postman docs.
- Updated the readme with the new Node.js requirements.

I'm including the readme file for the file uploading process below
## File Uploading In Deep Lynx
Users can now upload files into the Deep Lynx data warehouse. At time of writing (August 2020) files will either be stored on the local filesystem, or using Azure's Blob Storage service.

Users must configure the `FILE_STORAGE_METHOD` environment variable and set it to be either `azure_blob` or `filesystem`. Users must set additional environment variables with each of the methods, those are described inside the `.env-sample` file at the project's root.

Here is a brief diagram illustrating how file upload and storage works.
(DIAGRAM IS HERE)

1. User POSTS a multipart form to `/containers/:id/datasources/:id/files`. That multipart form can have numerous files and normal field/values attached.
2. Deep Lynx separates the files from the metadata and uploads them to the configured file storage.
3. Deep Lynx creates a record in the `files` table with the files location on the configured storage method and some metadata about the file.
4. If metadata (other fields) were present on the initial POST, Deep Lynx creates an "import" with a single record. That record contains the metadata and file names of all uploaded files. This information then goes through the normal data processing loop.
5. Users can query files directly from the `files` table or via metadata (once its gone through the processing loop)

### CODE REVIEW

File: src/api/data_source_routes.ts [hunk 1]
```diff
@@ -25,14 +29,18 @@ export default class DataSourceRoutes {
         app.delete("/containers/:id/import/datasources/:sourceID/active",...middleware, authInContainer("read", "data"),this.setInactive);
 
         app.get("/containers/:id/import/datasources/:sourceID/imports",...middleware, authInContainer("read", "data"),this.listDataSourcesImports);
-        app.post("/containers/:id/import/datasources/:sourceID/imports",...middleware, authInContainer("write", "data"),this.createManualJsonImport);
+        app.post("/containers/:id/import/datasources/:sourceID/imports",...middleware, fileUpload({limits:{fileSize: 50 * 1024 *1024}}), authInContainer("write", "data"),this.createManualJsonImport);
```
Comment (DnOberon) [inline src/api/data_source_routes.ts:32]: until we rewrite the json import using busboy we must use this middleware. I can't include this on the main router anymore as it interferes with `busboy`. Moved it from the main router to the route that needs it.

File: src/api/data_source_routes.ts [hunk 2]
```diff
@@ -295,5 +303,60 @@ export default class DataSourceRoutes {
...
+        const metadata: {[key: string]: any} = {}
+        let metadataFieldCount = 0;
+
+        // upload the file to the relevant file storage provider, saving the file name
+        // we can't actually wait on the full upload to finish, so there is no way we
+        // can take information about the upload and pass it later on in the busboy parsing
+        // because of this we're treating the file upload as fairly standalone
+        busboy.on('file', (fieldname: string, file: NodeJS.ReadableStream, filename: string, encoding: string, mimeType: string) => {
+            const user = req.user as UserT
+            DataSourceUploadFile(req.params.id, req.params.sourceID, user.id!, filename, encoding, mimeType, file as Readable)
```
Comment (DnOberon) [inline src/api/data_source_routes.ts:319]: `file as Readable` must be there as the type system doesn't recognize that `Readable` implements the `NodeJS.ReadableStream` interface correctly.

File: api_documentation/Core.swagger_collection.json [hunk 3]
```diff
@@ -1425,6 +1433,42 @@
...
+          {
+            "name": "dataSourceID",
+            "in": "path",
+            "required": true,
+            "type": "string",
+            "description": ""
+          }
+        ],
+        "responses": {
+          "200": {
```
Comment (ebohneyINL) [inline api_documentation/Core.swagger_collection.json:1465]: Do we want the response to be a 201, with it being a POST and a resource being created?

File: api_documentation/Core.swagger_collection.json [hunk 4]
```diff
@@ -1890,6 +1934,101 @@
         }
       }
     },
+    "/users/{user_id}/keys": {
+      "get": {
+        "description": "Retrieves only the active keys for supplied user. Secerts are hashed and are never revealed past initial creation.",
```
Comment (ebohneyINL) [inline api_documentation/Core.swagger_collection.json:1939]: Minor spelling: Secerts => Secrets

File: src/data_storage/file_storage.ts [hunk 5]
```diff
@@ -1,4 +1,4 @@
-import {FileT, FilesT, filesT} from "../types/fileT"
+import {FileT, FilesT, filesT, fileT} from "../types/fileT"
```
Comment (ebohneyINL) [inline src/data_storage/file_storage.ts:1]: Is there a need for FilesT any more on the import?

File: src/file_storage/file_storage.ts [hunk 6]
```diff
@@ -1,20 +1,40 @@
...
+    size: number // size in KB
+    metadata: {[key:string]: any} // adapter specific metadata if needed
+    adapter_name: string
+}
+
+// Returns an instantiated FileStorage provider if provider method is set, or user
+// can provide a specific adapter name to fetch an instance of specified adapter
+export default function FileStorageProvider(adapterName?: string): FileStorage | null {
+    switch ((adapterName) ? adapterName : Config.file_storage_method) {
         case "azure_blob_storage": {
```
Comment (ebohneyINL) [inline src/file_storage/file_storage.ts:26]: I'm not sure what the calls to this function look like, but the switch for the `adapterName` parameter has a case for "azure_blob_storage", but the actual implementation for the azure_blob file storage returns a name of "azure_blob". Is there a discrepancy there?

Comment (DnOberon) [reply src/file_storage/file_storage.ts:26]: Good catch, I'll correct that

File: src/data_storage/file_storage.ts (hunk 5 above)
Comment (DnOberon) [reply src/data_storage/file_storage.ts:1]: No, will correct

File: api_documentation/Core.swagger_collection.json (hunk 3 above)
Comment (DnOberon) [reply api_documentation/Core.swagger_collection.json:1465]: Not sure how to change this tbh. This is the swagger documentation which is autogenerated by the Postman documentation. Apart from manually editing the file, which I don't know what havoc that will cause, I don't know how to change it.

Comment (ebohneyINL) [reply api_documentation/Core.swagger_collection.json:1465]: Ahh, gotcha. There are ways with C#, I don't know how to do it with node.

**Pull request merged**